client/src/session/AuthContextProvider.tsx

```diff
@@ -0,0 +1,56 @@
+import React, {
+  createContext, ReactNode, useContext, useEffect, useState,
+} from 'react';
+import { useApi } from '../api/ApiContextProvider';
+import { isApiError } from '../api/ApiError';
+
+export interface AuthState {
+  authenticated: boolean
+  setAuthenticated: (newValue: boolean) => void
+}
+
+const AuthContext = createContext<AuthState | null>(null);
+
+export const AuthContextProvider = ({ children }: { children: ReactNode }) => {
+  const api = useApi();
```

Comment:
will there be a situation when on rerender this hook returns a new instance of `api` object ?

Reply:
There's not. I struggle with this one as to the best way to inject some sort of api object into the react code. Open to hearing what's worked well for you.

I've used global objects before but I've found them more difficult to mock in tests, at least the last time I tried it. Could have gotten better at mocking since then :).

Reply:
Below example will ensure that client instance is preserved during re-renders (it does not rerender in your case but it would if ContextProvider is placed inside BrowserRouter).

```tsx
export const AppContextProvider = ({ children }: { children: ReactNode }) => {
  const context = useMemo(() => new AppContext());
  return (
  <AppContext.Provider value={context}>
    {children}
  </AppContext.Provider>
);
```